src/api/server.rs

```diff
@@ -0,0 +1,134 @@
+use super::{handler, schema};
+use async_graphql::{
+    http::{playground_source, GraphQLPlaygroundConfig},
+    QueryBuilder,
+};
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::{IpAddr, SocketAddr};
+use tokio::sync::oneshot;
+use warp::{http::Response, Filter};
+
+pub struct Server {
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: oneshot::Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<oneshot::Receiver<()>>,
+
+    /// Enables the playground
+    playground: bool,
+}
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Self {
+        let (trigger_cancel, cancel_signal) = oneshot::channel::<()>();
+
+        Server {
+            address,
+            trigger_cancel,
+            cancel_signal: Some(cancel_signal),
+            playground: true,
+        }
+    }
+
+    pub fn set_playground(mut self, enable: bool) -> Self {
+        self.playground = enable;
+        self
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> IpAddr {
+        self.address.ip()
+    }
+
+    /// String representation of the bound port
+    pub fn port(&self) -> u16 {
+        self.address.port()
+    }
+
+    /// Stops the running API server
+    pub fn stop(self) {
+        let _ = self.trigger_cancel.send(());
+    }
+
+    /// Run the API server
+    pub async fn run(mut self) -> Self {
```

Comment:
It's usually good to separate the "initialization routine" from the "runtime routine".

I.e. have a way to do everything that's going on this fn and wait for it to complete - and have another fn to start the `tokio::spawn(server);`.

It would make sense to have separate structs as initialization routine arguments (i.e. `Config`) to avoid for the function to take `self` and return `self` - but the second `self` missing the `cancel_channel` and not suitable for calling `run` on it anymore.

Reply:
Thanks, makes sense. I'll work on improving this. I'd also like to get the global shutdown coordinator to signal closure, rather than rely on a separate channel. I'll play about with it some more and show you know what I come up with.

Reply:
You can take inspiration from how file server does it: https://github.com/timberio/vector/blob/1e180bff8f1635b7d24cd9f4838af63255ec5137/lib/file-source/src/file_server.rs#L68
It takes an abstract future from outside as a shutdown signal and terminates when the future resolves.